pinot-core/src/main/java/com/linkedin/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java

```diff
@@ -16,143 +16,602 @@
+                  hold();
+                }
+              }
+              break;
+            default:
+              segmentLogger.error("Holding after response from Controller: {}", response.toJsonString());
+              hold();
+              break;
+          }
+        }
+      } catch (Exception e) {
+        segmentLogger.error("Exception while in work", e);
+        _state = State.ERROR;
+      }
+    }
+  }
+
+  private File makeSegmentDirPath() {
+    return new File(_resourceDataDir, _segmentZKMetadata.getSegmentName());
+  }
+
+  private boolean buildSegment(boolean buildTgz) {
+    // Build a segment from in-memory rows.If buildTgz is true, then build the tar.gz file as well
+    File tempSegmentFolder = new File(_resourceTmpDir, "tmp-" + _segmentNameStr + "-" + String.valueOf(now()));
+
+    // lets convert the segment now
+    RealtimeSegmentConverter converter =
+        new RealtimeSegmentConverter(_realtimeSegment, tempSegmentFolder.getAbsolutePath(), _schema,
+            _segmentZKMetadata.getTableName(), _segmentZKMetadata.getSegmentName(), _sortedColumn, _invertedIndexColumns);
+
+    logStatistics();
+    segmentLogger.info("Trying to build segment");
+    final long buildStartTime = now();
+    try {
+      converter.build(_segmentVersion);
+    } catch (Exception e) {
+      segmentLogger.error("Could not build segment", e);
+      return false;
+    }
+    final long buildEndTime = now();
+    segmentLogger.info("Successfully built segment in {} ms", (buildEndTime - buildStartTime));
+    File destDir = makeSegmentDirPath();
+    FileUtils.deleteQuietly(destDir);
+    try {
+      FileUtils.moveDirectory(tempSegmentFolder.listFiles()[0], destDir);
+      if (buildTgz) {
+        TarGzCompressionUtils.createTarGzOfDirectory(destDir.getAbsolutePath());
+      }
+    } catch (IOException e) {
+      segmentLogger.error("Exception during move/tar segment", e);
+      FileUtils.deleteQuietly(tempSegmentFolder);
+      return false;
+    }
+    FileUtils.deleteQuietly(tempSegmentFolder);
+    return true;
+  }
+
+  private boolean commitSegment() {
+    // Send segmentCommit() to the controller
+    // if that succeeds, swap in-memory segment with the one built.
+    File destSeg = makeSegmentDirPath();
+    final String segTarFileName = destSeg.getAbsolutePath() + TarGzCompressionUtils.TAR_GZ_FILE_EXTENTION;
+    try {
+      _protocolHandler.segmentCommit(_currentOffset, _segmentNameStr, new File(segTarFileName));
+    } catch (FileNotFoundException e) {
+      segmentLogger.error("Tar file {} not found", segTarFileName, e);
+      return false;
+    }
+    return true;
+  }
+
+  private boolean buildSegmentAndReplace() {
+    boolean success = buildSegment(false);
+    if (!success) {
+      return success;
+    }
+    _realtimeTableDataManager.replaceLLSegment(_segmentZKMetadata.getSegmentName());
+    return false;
+  }
+
+  private void hold() {
+    try {
```

Comment:
th only reason this thread should be interrupted is when shutting down? (or, will thre be spurious interrupts)?
Let us discuss this some more